Add request generating and response parsing sections

xtlsheep opened the pull request.

xtlsheep committed:
Add request generating and response parsing sections

xtlsheep committed:
Update draft-ietf-protected-audience-key-value-services.md

xtlsheep committed:
Update draft-ietf-protected-audience-key-value-services.md

brusshamilton commented on draft-ietf-protected-audience-key-value-services.md:
```diff
@@ -318,9 +318,43 @@ Each key group is expected to have exactly one tag from the following list:
+1. Let `partitions` be an empty array.
+1. For each `signal` in `signals`:
+    1. Let `partition` be an empty map.
+    1. Set `partition["id"]` to 0.
```
Should we be setting `partition["id"]` to `compression group id` here? Otherwise all the partitions will have an ID of 0.

xtlsheep replied:
Gotcha, this request generation part will be updated to a new version that takes a list of compression group, which has a list of partitions, and use counters to assign compression group id and partition id. Will let you know after the new patch is uploaded.

brusshamilton commented on draft-ietf-protected-audience-key-value-services.md:
```diff
@@ -579,9 +613,51 @@ The output is a `response` to be sent to a Client.
+
+1. Use `request context` as the context to decrypt `encrypted response` and obtain `framed response`, returning failure if decryption fails.
+1. Remove and extract the first 5 bytes from `framed response` as the framing header (described in {{framing}}), removing them from `framed response`.
+1. If `framing header`'s `Version` field is not 0 or 2, return failure.
```
The framing header doesn't have a Version field. Do you mean "Compression"?

xtlsheep replied:
Yeah, changed it to `Compression`.

brusshamilton commented on draft-ietf-protected-audience-key-value-services.md:
```diff
@@ -579,9 +613,51 @@ The output is a `response` to be sent to a Client.
+1. For each `group` in `response["compressionGroups"]`:
+    1. If `group` is not a map, return failure.
+    1. If `group["content"]` does not exist, return failure.
+    1. If `framing header`'s `Version` field is 0:
```
You could probably combine these cases to something like:

Set `serialized content` to the result of decompressing `group["content"]` according to the compression algorithm specified in the `framing header`'s `Compression` field, returning failure if decompression fails.

xtlsheep replied:
Done.

brusshamilton commented on draft-ietf-protected-audience-key-value-services.md:
```diff
@@ -579,9 +613,51 @@ The output is a `response` to be sent to a Client.
+        1. If `partition` is not a map, return failure.
+        1. If `partition["keyGroupOutputs"]` does not exist, or is not an array, return failure.
+        1. Let `result` be an empty map.
+        1. For each `output` in `partition["keyGroupOutputs"]`:
```
You're not doing anything with the partition IDs or compression IDs here. I think those are necessary in order to map the response back to the request.

xtlsheep replied:
Added compression group id and partition id to the output result map.

xtlsheep committed:
Update request generating section

Revise the Request Generating section to utilize a new input format. The updated input represents a two-level map that organizes multiple partitions within multiple compression groups. This change aims to enhance compatibility with the W3C specification.

xtlsheep committed:
Update for comments

xtlsheep committed:
Update draft-ietf-protected-audience-key-value-services.md

brusshamilton commented on draft-ietf-protected-audience-key-value-services.md:
```diff
@@ -320,28 +320,30 @@ Each key group is expected to have exactly one tag from the following list:
+This section describes how the client MAY form and serialize request messages in order to fetch values from the Trusted Key Value server.
 
-This algorithm takes as input an [HPKE] `public key` and its associated `key id`, a `metadata` map for global configuration, where both keys and values are strings, and a list of `signals`, each of which is a map, from a list of strings to a list of strings.
+This algorithm takes as input an [HPKE] `public key` and its associated `key id`, a `metadata` map for global configuration, where both keys and values are strings, and a `compression groups` list, each of which is a map, from key `"partitions"` to value `partitions` as a list of maps.
```
I'm having trouble parsing what `compression groups` list looks like. What do you mean by "from key `"partitions"` to value `partitions`? I don't see a definition from either type, and I'm not sure what the quote characters in `"partitions`" indicate.

xtlsheep replied:
As discussed, updated the description of input arguments, and add a new field "namespace" in the partition struct.

xtlsheep committed:
Update draft-ietf-protected-audience-key-value-services.md

brusshamilton commented on draft-ietf-protected-audience-key-value-services.md:
```diff
@@ -320,28 +320,30 @@ Each key group is expected to have exactly one tag from the following list:
-    1. Set `partition["compressionGroupId"]` to `compression group id`.
-    1. Let `compression group id` increase by 1.
+1. For each `group` in `compression groups`:
+  1. For each `partition` in `compression groups["partitions"]:
```
You're looping over `compression groups` and then over a field in  `compression groups`. It seems like that would give you all combinations of both, which is probably not what you want.

xtlsheep replied:
As discussed, this is a typo here. Changed `compression groups` to `group`.

xtlsheep committed:
Update draft-ietf-protected-audience-key-value-services.md

brusshamilton commented on draft-ietf-protected-audience-key-value-services.md:
```diff
@@ -318,9 +318,43 @@ Each key group is expected to have exactly one tag from the following list:
+    1. Set `p["id"]` to `partition["id"]`.
+    1. Set `p["metadata"]` to `partition["metadata"].
+    1. Let `arguments` be an empty array.
+    1. For each `tag` → `value` in `partition["namespace"]`:
```
We better not copy each tag blindly. We should look through each tag in [2.2.3.1](https://privacysandbox.github.io/draft-ietf-protected-audience-key-value-service/draft-ietf-protected-audience-key-value-services.html#section-2.2.3.1) so that we only extract recognized tags. Otherwise it may be a problem if we add new tag types.

xtlsheep replied:
Add a if case to check if the tag is one of those available tags.

brusshamilton replied:
I still think it is cleaner to loop through the available tags and then see if the tag exists, but I guess this works as well.

brusshamilton commented on draft-ietf-protected-audience-key-value-services.md:
```diff
@@ -318,9 +318,43 @@ Each key group is expected to have exactly one tag from the following list:
+    1. Padding MAY be added to `framed payload`.
+    1. Return an empty `request` on failure of any of the previous steps.
+1. [HPKE] encrypt `framed payload` using `public key` and `key id` as in {{request-encryption}} to get the [HPKE] encrypted ciphertext `request` and [HPKE] encryption context `request context`.
+1. Return`request` and `request context`.
```
Return `request` and `request context`.
(add missing space)

xtlsheep replied:
Done

brusshamilton reviewed (commented):
Almost there, I think.

xtlsheep committed:
Update draft-ietf-protected-audience-key-value-services.md

xtlsheep committed:
Update draft-ietf-protected-audience-key-value-services.md

xtlsheep committed:
Update draft-ietf-protected-audience-key-value-services.md

lusayaa commented on draft-ietf-protected-audience-key-value-services.md:
```diff
@@ -579,9 +624,52 @@ The output is a `response` to be sent to a Client.
+a response from a Trusted Key Value service.
+
+It takes as input the `request context` returned from
+{{request-generate}} in addition to the `encrypted response`.
```
I wonder if you also want to indicate here that this is an example that only is looking to accept no compression or gzip (0 or 2), which is why it checks for 0 or 2. 

Or you may want to change the language from "Client MUST" to "Client MAY". Otherwise it makes it seem like all clients MUST reject compression bit 1.

xtlsheep replied:
But don't we want to reject the the responses with compression bit as 1 for now?

lusayaa replied:
Oh, for now, yes. We can update this later then.

JensenPaul commented on draft-ietf-protected-audience-key-value-services.md:
```diff
@@ -318,9 +318,54 @@ Each key group is expected to have exactly one tag from the following list:
+1. For each `group` in `compression groups`:
+  1. For each `partition` in `group["partitions"]:
+    1. Let `p` be an empty map.
+    1. Set `p["compressionGroupId"]` to `group[compressionGroupId"]`.
```
```suggestion
    1. Set `p["compressionGroupId"]` to `group["compressionGroupId"]`.
```

JensenPaul commented on draft-ietf-protected-audience-key-value-services.md:
```diff
@@ -318,9 +318,54 @@ Each key group is expected to have exactly one tag from the following list:
+    1. Set `p["metadata"]` to `partition["metadata"].
+    1. Let `arguments` be an empty array.
+    1. For each `tag` → `value` in `partition["namespace"]`:
+      1. If `tag` is one of [available tags](https://privacysandbox.github.io/draft-ietf-protected-audience-key-value-service/draft-ietf-protected-audience-key-value-services.html#section-2.2.3.1):
```
I don't think this should be an HTTP reference, but should instead be a reference to #tags or perhaps {{Available Tags}}

JensenPaul commented on draft-ietf-protected-audience-key-value-services.md:
```diff
@@ -318,9 +318,54 @@ Each key group is expected to have exactly one tag from the following list:
+`key id`, a `metadata` map for global configuration, where both keys and
+values are strings, and `compression groups`, which is a list of
+compression group structs. Each struct contains a compression group with
+a compression group ID and one or more partitions. Each partition
```
```suggestion
a `compressionGroupID` and one or more `partition`s. Each `partition`
```
I'm not sure if I'm using the right quotes, but I think we need to make this match the steps below

xtlsheep replied:
Changed to `compression group id` which represents the value of partition["compressionGroupId"].

JensenPaul commented on draft-ietf-protected-audience-key-value-services.md:
```diff
@@ -318,9 +318,54 @@ Each key group is expected to have exactly one tag from the following list:
+values are strings, and `compression groups`, which is a list of
+compression group structs. Each struct contains a compression group with
+a compression group ID and one or more partitions. Each partition
+includes multiple keys, such as "id," whose value is an integer;
```
I don't think we should say "such as" if they're required, which according to the steps below it looks like they are.

xtlsheep replied:
Changed to "which contain"

JensenPaul commented on draft-ietf-protected-audience-key-value-services.md:
```diff
@@ -318,9 +318,54 @@ Each key group is expected to have exactly one tag from the following list:
+This algorithm takes as input an [HPKE] `public key` and its associated
+`key id`, a `metadata` map for global configuration, where both keys and
+values are strings, and `compression groups`, which is a list of
+compression group structs. Each struct contains a compression group with
```
```suggestion
compression group structs. Each compression group struct contains
```

JensenPaul commented on draft-ietf-protected-audience-key-value-services.md:
```diff
@@ -318,9 +318,54 @@ Each key group is expected to have exactly one tag from the following list:
+values are strings, and `compression groups`, which is a list of
+compression group structs. Each struct contains a compression group with
+a compression group ID and one or more partitions. Each partition
+includes multiple keys, such as "id," whose value is an integer;
```
I don't think we should say "includes multiple keys" as this is not very definitive or specific.  I'd say it's a struct with some required fields, or it's a map with specific key and value types.

xtlsheep replied:
Changed this sentence to
```
Each `partition` is a map with specified keys, which are "id," ...
```

JensenPaul commented on draft-ietf-protected-audience-key-value-services.md:
```diff
@@ -318,9 +318,54 @@ Each key group is expected to have exactly one tag from the following list:
+compression group structs. Each struct contains a compression group with
+a compression group ID and one or more partitions. Each partition
+includes multiple keys, such as "id," whose value is an integer;
+"metadata," whose value is a map where both keys and values are strings;
+and "namespace," whose value is a map where keys are strings and values
```
```suggestion
"metadata", whose value is a map where both keys and values are strings;
and "namespace", whose value is a map where keys are strings and values
```

JensenPaul commented on draft-ietf-protected-audience-key-value-services.md:
```diff
@@ -579,9 +624,52 @@ The output is a `response` to be sent to a Client.
+This section describes how a conforming Client MUST parse and validate
+a response from a Trusted Key Value service.
+
+It takes as input the `request context` returned from
+{{request-generate}} in addition to the `encrypted response`.
```
```suggestion
It takes as input the `request context`, returned from
{{request-generate}}, and an `encrypted response`, an array of bytes from the Key Value Service generated by {{Generating a Response}}.
```
I don't know if I'm using the right syntax here.

xtlsheep committed:
Update draft-ietf-protected-audience-key-value-services.md

xtlsheep committed:
Update draft-ietf-protected-audience-key-value-services.md

xtlsheep committed:
Update draft-ietf-protected-audience-key-value-services.md

JensenPaul commented on draft-ietf-protected-audience-key-value-services.md:
```diff
@@ -318,9 +318,57 @@ Each key group is expected to have exactly one tag from the following list:
+This section describes how the client MAY form and serialize request
+messages in order to fetch values from the Trusted Key Value server.
+
+This algorithm takes as input an [HPKE] `public key` and its associated
+`key id`, a `metadata` map for global configuration, where both keys and
+values are strings, and `compression groups`, which is a list of `group`s
+with the following parameters:
+* `compressionGroupId`, an integer indicates the index of this compression
+  group.
+* `partitions`, contains all the `partition`s belong to this compression
+  group. A single `partition` has the following parameters:
+  * `id`, an integer indicates the index of this partition.
+  * `namespace`, an empty map, whose keys are strings and values are list of
+    strings.
+  * `metadata`, an empty map, whose keys and values are strings.
```
```suggestion
This algorithm takes as input:
* an [HPKE] `public key`.
* a `key id` integer associated with `public key`.
* a `metadata` map for global configuration, where both keys and
values are strings.
* a `compression groups`, which is a list of `group`s, each
with the following parameters:
 * a `compressionGroupId` integer identifier of this compression
   group.
 * a `partitions`, which is a list of `partition`s belonging to this compression
   group. Each `partition` has the following parameters:
   * an `id` integer identifier of this partition.
   * a `namespace` map whose keys are strings and values are list of
     strings.
   * a `metadata` map whose keys and values are strings.
```

xtlsheep committed:
Update draft-ietf-protected-audience-key-value-services.md

Co-authored-by: Paul Jensen <JensenPaul@users.noreply.github.com>
